Add rust-libp2p v0.52 blog post (#1)
mxinden/blog: src/_blog/rust-libp2p-052-release.md

---
tags:
- libp2p
- rust
title: rust-libp2p v0.52 release spotlight
description: Walking through the rust-libp2p v0.52 release
date: 2023-06-19
permalink: "/2023-06-19-rust-libp2p-052"
translationKey: ''
header_image: /rust-libp2p-2022-header.png
author: Thomas Eizinger
---

# rust-libp2p v0.52 release spotlight

> With the `v0.52` release, the `rust-libp2p` maintainer team tried something new.

Author: I find this paragraph difficult to read. "tried something new. Breaking changes" sounds like we re-invented the idea of breaking changes.

The below makes it a bit more explicit. Though I would advocate for moving the entire paragraph towards the end of the section.

```suggestion
With the `v0.52` release, the `rust-libp2p` maintainer team tried something new - holding back breaking changes, releasing as much as possible in a non-breaking way.
```

Reviewer: Yeah I am not particularly happy with the introduction actually. I want to make these posts a recurring thing so maybe it is better to just dive right in and talk about the changes?

Author: Diving right in sounds good to me.

Breaking changes - even in the presence of an obvious or easy upgrade path - are a burden for our users.
They easily trigger a cascade of breaking changes for downstream users.

Coming up to the `v0.52` release, we consciously held back and batched up breaking changes and released several features as patch releases instead.
As a result, `v0.52` is unfortunately quite large.

> This blogpost aims to highlight some of the more exciting changes:

Author: ```suggestion
This blogpost highlights the most exciting changes:
```

Not a native speaker. To me "the more exciting changes" sounds like the other non-highlighted changes are bad.

- Automatic kademlia client/server mode
- Type-safe `/p2p` multiaddresses
- More consistent event/command naming
- Improve ergonomics around stream-upgrade errors

> - Simpler noise interface

Author: In case you want to add more things to the list. Totally fine with keeping it as is.

https://github.com/libp2p/rust-libp2p/pull/3952

Reviewer: Yeah that is a good one actually!

  
    /// The local [`ConnectionHandler`] added or removed support for one or more protocols.
    LocalProtocolsChange(ProtocolsChange<'a>),
    /// The remote [`ConnectionHandler`] now supports a different set of protocols.
    RemoteProtocolsChange(ProtocolsChange<'a>),
  }
  
  pub enum ProtocolsChange<'a> {
    Added(ProtocolsAdded<'a>),
    Removed(ProtocolsRemoved<'a>),
  }
  ``` 

  `ProtocolsAdded` and `ProtocolsRemoved` are iterators over a (new) type called `StreamProtocol`.
- Protocols are now enforced to be valid UTF-8 strings: [#3746](https://github.com/libp2p/rust-libp2p/pull/3746).

  This was always the case in the specs but the `rust-libp2p` implementation was lagging behind here and improperly represented them as bytes internally.
- Local changes to our protocols (i.e. a node going from Kademlia server to client mode) are now immediately pushed to the remote via the [`/ipfs/id/push/1.0.0`](https://github.com/libp2p/specs/tree/master/identify#identifypush) protocol: [#3980](https://github.com/libp2p/rust-libp2p/pull/3980).
- Simplify the scoring mechanism of external addresses: [#3954](https://github.com/libp2p/rust-libp2p/pull/3954)

> Not only does this work out-of-the-box and thus improves the developer experience of `rust-libp2p`, it should also result in a much more useful and up-to-date routing table for all users of the Kademlia DHT.

Author: There are many libp2p Kademlia DHT deployments out there. I find talking of THE Kademlia DHT confusing.

```suggestion
Not only does this work out-of-the-box and thus improves the developer experience of `rust-libp2p`, it should also result in a much more useful and up-to-date routing table for all nodes on a Kademlia DHT.
```

## Type-safe `/p2p` multiaddresses

The `/p2p` protocol of multiaddresses specifies the identity of a peer in the form of a `PeerId` such as `12D3KooWETLZBFBfkzvH3BQEtA1TJZPmjb4a18ss5TpwNU7DHDX6`.
Yet for the longest time, the type-definition of the `/p2p` protocol looked like this:

```rust
pub enum Protocol<'a> {
    // omitted other variants ...
    P2p(Multihash),
}
```

Every `PeerId` is a valid `Multihash` but not vice-versa.
Dang!
That is a lot of "impossible" errors that the type-system should avoid.

> [Thanks](https://github.com/multiformats/rust-multihash/issues/322) [to](https://github.com/libp2p/rust-libp2p/pull/3514) [a](https://github.com/libp2p/rust-libp2p/pull/3656) [lot](https://github.com/libp2p/rust-libp2p/pull/3350) [of](https://github.com/multiformats/rust-multihash/pull/272) [work](https://github.com/multiformats/rust-multiaddr/pull/83), it now looks like this:

Author: :D
